File: src/main/java/org/adoptopenjdk/jheappo/objects/PrimitiveArray.java
package org.adoptopenjdk.jheappo.objects;

/*
 * Copyright (c) 2018 Kirk Pepperdine.
 * Licensed under https://github.com/AdoptOpenJDK/jheappo/blob/master/LICENSE
 * Instructions: https://github.com/AdoptOpenJDK/jheappo/wiki
 */

import org.adoptopenjdk.jheappo.heapdump.EncodedChunk;
import org.adoptopenjdk.jheappo.model.BasicDataTypeValue;


public class PrimitiveArray extends HeapObject {

    public static final int TAG = 0x23;

    private int stackTraceSerialNumber;
    private int size;
    private byte elementType;
    private char signature = ' ';
    BasicDataTypeValue[] elements;

    public PrimitiveArray(EncodedChunk buffer) {
        super(buffer);
        stackTraceSerialNumber = buffer.extractInt();
        size = buffer.extractInt();
        elementType = buffer.extractByte();
        readArray(buffer,elementType,size);
    }

    private void readArray(EncodedChunk buffer, byte elementType, int size) {
        BasicDataTypes dataType = BasicDataTypes.fromInt(elementType);
        signature = BasicDataTypes.fromInt(elementType).getMnemonic();
        if ( dataType.equals(BasicDataTypes.UNKNOWN)) {
            throw new IllegalArgumentException("Unknown data type : " + elementType);
        }
        elements = new BasicDataTypeValue[size];
        for (int i = 0; i < size; i++) {
            elements[i] = buffer.extractBasicType(dataType);
        }
    }
}

File: src/main/java/org/adoptopenjdk/jheappo/objects/PrimitiveArray.kt
package org.adoptopenjdk.jheappo.objects

/*
 * Copyright (c) 2018 Kirk Pepperdine.
 * Licensed under https://github.com/AdoptOpenJDK/jheappo/blob/master/LICENSE
 * Instructions: https://github.com/AdoptOpenJDK/jheappo/wiki
 */

import org.adoptopenjdk.jheappo.heapdump.EncodedChunk
import org.adoptopenjdk.jheappo.model.BasicDataTypeValue


class PrimitiveArray(buffer: EncodedChunk) : HeapObject(buffer) {

    companion object {
        const val TAG = 0x23
    }

    private val stackTraceSerialNumber: Int = buffer.extractInt()
    private val size: Int = buffer.extractInt()
    private val elementType: Byte = buffer.extractByte()
    private val signature: Char

    internal val elements: Array<BasicDataTypeValue>

    init {
        val dataType = BasicDataTypes.fromInt(elementType.toInt()) ?: BasicDataTypes.UNKNOWN
        if (dataType == BasicDataTypes.UNKNOWN) {
            throw IllegalArgumentException("Unknown data type : $elementType")
        }
        signature = BasicDataTypes.fromInt(elementType.toInt())!!.mnemonic
        elements = (0 until size).map { buffer.extractBasicType(dataType) }.toTypedArray()
    }
}
